feat: 홈 투표 UI 및 내비게이션 바 구현 (#16)

rafa-e1 opened the pull request:
# 🚀 Pull Request

> **No merge conflict? Where’s the fun in that.**

<br>

## 📌 요약

<!-- 어떤 변경이 이루어졌는지 간단히 작성해주세요 -->
- 커스텀 내비게이션 바 UI 구현
- HomeSection 기반의 Compositional Layout 분리
- SectionController 기반 UI 아키텍처 도입
- “전체보기” 버튼 동작 ViewModel로 위임 처리

<br>

## 📄 상세 설명

<!-- 배경, 의도, 고려한 사항 등을 자세히 작성해주세요 -->
- 각 섹션별 레이아웃을 HomeLayoutFactory를 통해 추상화하여 구성 유연성을 확보함
- VoteSectionLayout, DefaultSectionLayout을 구현하여 섹션별 디자인 차이를 명확히 분리
- 셀, 헤더, 푸터 바인딩 책임을 SectionController 프로토콜로 위임하여 ViewController 로직 단순화
- HomeSectionFooterView의 “전체보기” 버튼 탭 시 해당 섹션 타이틀을 ViewModel로 전달하도록 로직 연결
- 디렉토리 구조는 다음과 같음
```
Home/
├── Component/
│   ├── Cell/
│   ├── Footer/
│   └── Header/
├── Enum/
├── Layout/
│   ├── Factory/
│   ├── Protocol/
│   └── Section/
│       ├── Controller/
│       ├── SectionLayout/
│       └── Supplementary/
├── View/
├── ViewModel/
```
## 폴더별 설명

### `Component/`
UI 컴포넌트 단위로 구성된 View 계층입니다.  
- `Cell/` : 각 섹션에 대응되는 `UICollectionViewCell`
- `Header/`, `Footer/` : `UICollectionReusableView` 형태로 구성되는 Supplementary View

### `Enum/`
- `HomeSection.swift`: 섹션 타입 정의와 타이틀/버튼 텍스트/identifier 등 섹션별 메타데이터 캡슐화

### `Layout/`

#### `Layout/Factory/`
- `HomeLayoutFactory.swift`: `HomeSection`에 따라 알맞은 레이아웃을 반환하는 팩토리

#### `Layout/Protocol/`
- `HomeSectionLayoutFactory`, `SectionLayoutProviding`: 레이아웃 생성을 추상화하는 인터페이스

#### `Layout/Section/`

##### `Layout/Section/Controller/`
- `SectionController.swift`: 셀 및 SupplementaryView의 구성 책임을 갖는 인터페이스
- `DefaultSectionController.swift`, `VoteSectionController.swift`: 각 섹션별 구성 로직 구현

##### `Layout/Section/SectionLayout/`
- `DefaultSectionLayout.swift`, `VoteSectionLayout.swift`: 섹션별 `NSCollectionLayoutSection` 생성 클래스

##### `Layout/Section/Supplementary/`
- `HomeSupplementaryItemFactory.swift`: Header/Footer SupplementaryItem 레이아웃 제공 유틸

### `View/`
- `HomeViewController.swift`: 컬렉션 뷰 셋업, 데이터 바인딩, 섹션 컨트롤러 호출 등 화면 전체를 조립

### `ViewModel/`
- `HomeViewModel.swift`: UI 이벤트 처리 및 화면 상태에 필요한 데이터를 가공

## 설계 전략 요약

| 요소 | 전략 |
|------|------|
| **UI 구성 책임 분리** | SectionController 패턴으로 셀/헤더/푸터 구성 분리 |
| **레이아웃 구성 책임 분리** | Factory 및 LayoutProvider 추상화 적용 |
| **확장성** | 섹션별 로직/레이아웃 추가 시 새로운 SectionController + Layout만 정의하면 됨 |
| **MVVM 적용** | View ↔ ViewModel 간 이벤트 처리 책임 분리 |
| **재사용성** | SupplementaryItem, Layout 구조는 다른 화면에서도 쉽게 재사용 가능 |

## 주요 코드 설명
### `HomeSectionLayoutFactory`
```swift
protocol HomeSectionLayoutFactory {
    func layout(for section: HomeSection) -> NSCollectionLayoutSection
}
```
- 역할: `HomeSection`에 해당하는 레이아웃을 생성해주는 팩토리의 인터페이스입니다.
- 사용 이유: 섹션에 따라 레이아웃이 달라지기 때문에 이를 동적으로 선택하고 추상화하여 깔끔한 분기 구조를 유지하기 위해 사용합니다.

### `SectionLayoutProviding`
```swift
protocol SectionLayoutProviding {
    func layout() -> NSCollectionLayoutSection
}
```
- 역할: 레이아웃을 제공하는 책임만 가진 인터페이스입니다.
- 사용 이유: 공통된 레이아웃 반환 방식(`layout()`)을 통해 다양한 레이아웃을 하나의 타입으로 묶기 위해 사용합니다.

### `DefaultSectionLayout`
```swift
final class DefaultSectionLayout: SectionLayoutProviding {
    func layout() -> NSCollectionLayoutSection { ... }
}
```
- 역할: 주변 맛집, 추천, 찐 맛집 등과 같은 일반적인 섹션들의 레이아웃을 구성합니다.
- 특징: 셀 크기를 동일하게 유지하고 헤더와 푸터를 함께 포함하는 `NSCollectionLayoutSection`을 반환합니다.

### `VoteSectionLayout`
```swift
final class VoteSectionLayout: SectionLayoutProviding {
    func layout() -> NSCollectionLayoutSection { ... }
}
```
- 역할: 투표 섹션 전용 레이아웃을 구성합니다.
- 특징: `SupplementaryView`를 포함하지 않습니다.

### `HomeSupplementaryItemFactory`
```swift
enum HeaderSupplementaryItemFactory { static func create() -> NSCollectionLayoutBoundarySupplementaryItem }
enum FooterSupplementaryItemFactory { static func create() -> NSCollectionLayoutBoundarySupplementaryItem }
```
- 역할: 헤더와 푸터용 `SupplementaryItem` 레이아웃을 생성합니다.
- 사용 이유: 레이아웃 코드 중복을 줄이고 헤더/푸터의 높이와 정렬을 통일성 있게 유지하기 위함입니다.

### `SectionController`
```swift
protocol SectionController {
    func numberOfItems() -> Int
    func cellForItem(at:in:) -> UICollectionViewCell
    func header(in:at:) -> UICollectionReusableView?
    func footer(in:at:) -> UICollectionReusableView?
}
```
- 역할: 각 섹션의 셀, 헤더, 푸터를 구성하는 책임을 가진 인터페이스입니다.
- 사용 이유: `UICollectionViewDataSource` 구현을 분산시켜 `ViewController`를 단순화하고 섹션별 책임을 분리하기 위해 사용합니다.

### `DefaultSectionController`
```swift
final class DefaultSectionController: SectionController {
    // 생략
}
```
- 역할: 주변 맛집, 추천, 찐 맛집과 같은 일반 섹션의 셀/헤더/푸터를 구성하는 클래스입니다.
- 특징: `HomeSection`을 기반으로 `셀 identifier`, `타이틀`, `SupplementaryView` 여부 등을 참조하여 UI를 구성합니다.
- 사용 이유: 하나의 클래스로 다양한 섹션에 공통된 UI 패턴을 재사용할 수 있습니다.

### `VoteSectionController`
```swift
final class VoteSectionController: SectionController {
    // 생략
}
```
- 역할: 투표 섹션 전용 셀을 구성하는 클래스로, `SupplementaryView`는 사용하지 않습니다.
- 특징: `UICollectionViewCell` 하나만 구성하며 가장 간결한 섹션 컨트롤러의 예시입니다.
- 사용 이유: 레이아웃뿐 아니라 UI 구성 책임도 전용 클래스로 분리하여 역할을 명확히 하기 위함입니다.

### `HomeLayoutFactory`
```swift
final class HomeLayoutFactory: HomeSectionLayoutFactory {
    func layout(for section: HomeSection) -> NSCollectionLayoutSection { ... }
}
```
- 역할: 각 `HomeSection`에 대해 적절한 `SectionLayoutProviding` 구현체를 선택하여 `NSCollectionLayoutSection`을 반환하는 팩토리입니다.
- 사용 이유: `UICollectionViewCompositionalLayout`의 섹션별 구성을 `switch` 문으로 깔끔하게 위임하고 변경에 유연하게 대응할 수 있도록 하기 위함입니다.

<br>

## 📸 구현 화면 (GIF / Screenshot)

<!-- 동작 화면을 시뮬레이터 캡처나 GIF로 첨부해주세요. (Drag & Drop 또는 링크 삽입 가능) -->
| 구현 내용 | SE | 13 mini | 15 pro |
| :-------------: | :----------: | :----------: | :----------: |
| 홈 투표 UI | <img src = "https://github.com/user-attachments/assets/aee350a1-2d35-45e6-852a-fb00636f608f" width ="250"> | <img src = "https://github.com/user-attachments/assets/f5a5c44e-479d-42b3-bf53-22fa19cb635b" width ="250"> | <img src = "https://github.com/user-attachments/assets/e116659c-0bfe-4cbd-bc3b-053bd72129f5" width ="250"> |

<br>

## 🙋 리뷰 가이드

<!-- 리뷰어가 참고했으면 하는 부분, 중점적으로 봐야 할 로직이 있다면 적어주세요. -->
- SectionController 설계가 과도하게 추상화되지 않았는지 확인 부탁드립니다.

<br>

## 🔗 관련 이슈 (Issue)

<!-- 관련된 이슈 번호가 있다면 명시해주세요. (예: #123) -->
- Resolved: #16 

<br>

## ✅ PR Check List
- [x] Merge 대상 브랜치가 올바른가?
- [x] 최종 코드가 에러 없이 잘 동작하는가?
- [ ] 전체 변경사항이 500줄을 넘지 않는가?

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **신규 기능**
  * 홈 화면이 새롭게 도입되어 투표, 근처 랭킹, 개인 추천, 숨은 맛집 등 다양한 섹션이 추가되었습니다.
  * 홈 화면 내 각 섹션별로 커스텀 셀, 헤더/푸터 뷰, 내비게이션 바 등이 적용되었습니다.
  * 투표, 추천, 랭킹 등 다양한 데이터 모델과 뷰모델이 추가되어 정보를 효과적으로 표시합니다.
  * 커스텀 폰트, 색상, 아이콘 및 이미지 리소스가 추가되어 디자인 시스템이 확장되었습니다.
  * StackView 및 UIView 등 UI 편의 확장 기능이 도입되었습니다.

* **스타일**
  * UI 요소별 배경색, 폰트, 레이아웃 등 일관된 스타일이 적용되었습니다.

* **기타**
  * 불필요한 기존 HomeViewController 파일이 제거되었습니다.
  * Info.plist에 커스텀 폰트 리소스가 추가되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

rafa-e1 committed:
add: 로고 이미지 추가

rafa-e1 committed:
feat: 이미지 리소스 관리를 위한 열거형(MatzipImage) 구현

rafa-e1 committed:
add: 검색 아이콘 추가

rafa-e1 committed:
feat: 검색 아이콘 열거형 케이스 구현

rafa-e1 committed:
feat: 서브 텍스트 및 구분선 컬러 정의 추가

rafa-e1 committed:
feat: UIView extension을 통해 addSubviews 메서드 구현

rafa-e1 committed:
feat: 커스텀 내비게이션 바 구현

rafa-e1 committed:
feat: HomeViewModel 추가 및 내비게이션 바 구성 로직 구현

rafa-e1 committed:
feat: 뷰모델 프로토콜 기반 의존성 주입 및 커스텀 내비게이션 바 적용

rafa-e1 committed:
refactor: MainTab에서 HomeViewController에 뷰모델 의존성 주입 적용

rafa-e1 committed:
refactor: super.setupStyles() 호출로 메인 배경색 설정 로직 반영

rafa-e1 committed:
add: Pretendard 폰트 추가

rafa-e1 committed:
feat: 커스텀 폰트 적용을 위한 UIFont 확장자 및 FontName enum 구현

rafa-e1 committed:
feat: 공통 UICollectionView 기반 Base 뷰 클래스 구현 및 뷰 설정 메서드 추가

rafa-e1 committed:
add: 아이콘 및 이미지 추가

rafa-e1 committed:
add: 임시로 사용할 이미지 추가

rafa-e1 committed:
feat: 추가된 아이콘 및 이미지 케이스 구현

rafa-e1 committed:
feat: 신규 컬러 4종 추가

rafa-e1 committed:
rename: UIView+의 폴더명을 UI로 수정

- Common/Extension -> UI/Extension

rafa-e1 committed:
feat: UIStackView에 configureStackView 유틸 함수 추가하여 반복 코드 제거

rafa-e1 committed:
design: 커스텀 폰트 적용

rafa-e1 committed:
refactor: configureStackView 유틸 함수 적용

rafa-e1 committed:
feat: UIView 기반의 ReusableIdentifier 프로토콜 추가하여 identifier 자동화

rafa-e1 committed:
feat: 공통 문구 관리를 위한 MatzipText enum 추가

rafa-e1 committed:
feat: 투표 섹션 UI 구현

rafa-e1 committed:
add: 나머지 섹션 컬렉션 뷰 셀 파일 추가

- 주변 맛집 랭킹 셀
- 맞춤형 추천 셀
- 숨은 찐 맛집 셀

rafa-e1 committed:
fix: 레이아웃 오류 해결

rafa-e1 committed:
feat: HomeSection enum에 헤더/푸터 타이틀 및 셀 identifier 추상화 추가

rafa-e1 committed:
feat: 홈 섹션 헤더 및 푸터 UI 구현

rafa-e1 committed:
refactor: 타입 추가

rafa-e1 committed:
refactor: ViewModel 폴더 생성 및 HomeViewModel 추가

rafa-e1 committed:
feat: 홈 섹션 레이아웃 추상화 구조 설계

- HomeSectionLayoutFactory 프로토콜 정의

rafa-e1 committed:
feat: 공통 섹션 레이아웃 생성을 위한 SectionLayoutProviding 프로토콜 정의

rafa-e1 committed:
feat: 헤더 및 푸터 SupplementaryItem 생성을 위한 Factory 유틸 추가

rafa-e1 committed:
feat: 공통 섹션 레이아웃 생성을 위한 DefaultSectionLayout 구현

rafa-e1 committed:
feat: 투표 섹션 전용 Compositional Layout 구성용 VoteSectionLayout 추가

rafa-e1 committed:
feat: UICollectionView 섹션 구성 모듈화를 위한 SectionController 인터페이스 추가

rafa-e1 committed:
feat: 홈 섹션별 셀 및 SupplementaryView 관리를 위한 DefaultSectionController 구현

rafa-e1 committed:
feat: Header/Footer 없이 Vote 전용 셀만 구성하는 VoteSectionController 구현

rafa-e1 committed:
feat: Home 화면의 섹션별 레이아웃 생성을 담당하는 HomeLayoutFactory 구현

rafa-e1 committed:
feat: 홈 투표 섹션 추가 및 컬렉션 뷰 UI 구현

rafa-e1 committed:
chore: Xcode 프로젝트 메타데이터 업데이트 (LastUpgradeCheck 1640으로 변경)

rafa-e1 committed:
refactor: Supplementary 폴더를 Section 폴더 내부로 이동

rafa-e1 committed:
refactor: SectionLayout으로 폴더명 수정

- 기존에는 Layout/Section/Layout으로 구성되어 있어 폴더명이 중첩되어 SectionLayout으로 폴더명 변경

alstjr7437 commented on MatzipBook/MatzipBook/Core/DesignSystem/Foundation/MatzipImage.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+enum MatzipImage: String {
+    case imgLogo = "img_logo"
+    case imgBubble = "img_bubble"
+    case imgVS = "img_vs"
+}
```
사용되는 image, Text 들을 enum으로 관리해서 좋네요!!

혹시 따로 궁금한점인데 img_logo나 ic_clock등 snake case를 쓰신 이유가 궁금합니다!

rafa-e1 replied:
Asset Catalog에서는 이름 구분할 때 대소문자를 구분하지 않아서 이미지 등록할 때 스네이크 케이스로 등록하는 것이 충돌 가능성을 방지하는걸로 알고 있습니다!

alstjr7437 commented on MatzipBook/MatzipBook/Core/DesignSystem/Foundation/UIFont+.swift:
```diff
@@ -0,0 +1,20 @@
+    case regular = "Pretendard-Regular"
+}
+
+extension UIFont {
+    
+    static func applyFont(_ name: FontName, ofSize size: CGFloat) -> UIFont {
+        return UIFont(name: name.rawValue, size: size) ?? .systemFont(ofSize: size)
+    }
+}
```
오 폰트를 이렇게 쓰는 방법도 있군요 
저는 폰트를 쓸때 SwiftUI긴 하지만 
```swift
static var pretendardBold14: Font { .pretend(type: .bold, size: 14) }
static var pretendardBold16: Font { .pretend(type: .bold, size: 16) }
```

이런식으로 썻는데 범수님 같이 쓰는게 더 확장성이 좋아보이네요!!

rafa-e1 replied:
말씀해주신 방식도 충분히 좋은 접근이고 상황에 따라 더 적합할 수도 있다고 생각합니다!
결국 코드에는 정답은 없고 프로젝트의 성격과 팀의 방향성에 따라 유연하게 선택하는게 가장 중요하다고 생각해요

alstjr7437 commented on MatzipBook/MatzipBook/Core/UI/Extension/UIView+.swift:
```diff
@@ -0,0 +1,15 @@
+
+import UIKit
+
+extension UIView {
+    
+    func addSubviews(_ views: UIView...) {
+        views.forEach { addSubview($0) }
+    }
+}
```
이부분은 addSubview를 여러번 말고 한번에 다 추가하는 방식이네요!! 배워갑니다!! 🙏

...으로 list가 아닌 방식으로 하는 방식도 좋네요!

rafa-e1 replied:
감사합니다! 🙏
개인적으로 `...` 을 사용하면 줄바꿈했을 때 더 깔끔한 것 같더라고요:)

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Component/Cell/VoteCollectionViewCell.swift:
```diff
@@ -0,0 +1,217 @@
+        configureBubbleImageView(leftNameImageView)
+        configureBubbleImageView(rightNameImageView)
+        
+        configureLabel(
+            leftNameLabel,
+            text: "그집짬뽕0927",
+            fontName: .bold,
```
이 Text 같은 경우는 매개변수로 바꾸실 예정인거죠?

rafa-e1 replied:
현재는 UI 구성 상태만 확인하기 위해 임시로 텍스트를 넣어둔 상태이고 실제 가게명은 API 응답 데이터를 통해 받아와 뷰모델에서 바인딩할 예정입니다! 그래서 이 부분은 이후 네트워크 연결 및 뷰모델 설계 시 동적으로 처리될 계획입니다

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Component/Footer/HomeSectionFooterView.swift:
```diff
@@ -0,0 +1,90 @@
+    }
+    
+    override func setupConstraints() {
+        seeAllButton.snp.makeConstraints {
+            $0.centerY.horizontalEdges.equalToSuperview()
+            $0.height.equalTo(50)
+        }
```
아직 snapkit에 익숙하지 않아서 열심히 배우겠습니다!

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Component/HomeNavigationBarView.swift:
```diff
@@ -0,0 +1,98 @@
+    private func setupStyles() {
+        backgroundColor = .mainBackgroundColor
+
+        logoImageView.do {
+            $0.image = .imgLogo
+            $0.contentMode = .scaleAspectFit
+        }
```
Then을 쓰면 이렇게 간결해지는군요

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Enum/HomeSection.swift:
```diff
@@ -0,0 +1,50 @@
+//
+
+import Foundation
+
+enum HomeSection: Int, CaseIterable {
+    case vote, nearbyRanking, recommendation, hiddenGem
+    
+    var headerTitle: String? {
+        switch self {
+        case .vote: return nil
+        case .nearbyRanking: return "주변 맛집 랭킹"
```
상수값 들도 관리하는 것 같은데 HomeSection이라는 변수명으로 지으신 이유가 궁금해요!

rafa-e1 replied:
혹시 질문 주신 부분이 headerTitle이나 footerButtonTitle처럼 HomeSection 내부에 정의된 텍스트 관련 프로퍼티를 말씀하신 걸까요?

만약 그 부분을 말씀하신 거라면 HomeSection은 단순히 텍스트 상수를 모아놓은 enum이라기보다는, 홈 화면 내 각 섹션의 성격과 역할을 표현하고 관련된 뷰 구성 요소들(헤더, 푸터, 셀 ID, supplementary 여부 등)을 응집시킨 모델 타입으로 구성했습니다

혹시 제가 질문 의도를 잘못 이해했다면 다시 한번 말씀 부탁드립니다!

+ 지금 다시 코드를 보니 MVVM 패턴의 확장성 측면에서는 구조적으로 제한적일 수 있다는 것을 깨달았네요. 현재는 UI 구성 중심으로 빠르게 작업하면서 뷰모델 분리까지는 고려하지 못했는데 실제 데이터를 바인딩하거나 뷰 로직이 복잡해지는 시점에는 각 섹션을 뷰모델 단위로 분리하고 역할을 재정의하는 방향으로 구조를 개선해야겠어요,

alstjr7437 replied:
아하 각 세션의 성격을 표현하는 부분이였군요 

저가 물었던 부분은 이 부분이 상수값들을 모아둔 enum이랑 분리되는게 좋지 않을까 했는데

섹션의 성격이 저장된 부분이면 지금처럼 상수값이 들어있어도 괜찮을 것 같아요

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Layout/Section/Controller/SectionController.swift:
```diff
@@ -0,0 +1,27 @@
+
+import UIKit
+
+protocol SectionController {
+    func numberOfItems() -> Int
+    
+    func cellForItem(
+        at indexPath: IndexPath,
+        in collectionView: UICollectionView
+    ) -> UICollectionViewCell
+    
+    func header(
+        in collectionView: UICollectionView,
+        at indexPath: IndexPath
+    ) -> UICollectionReusableView?
+    
+    func footer(
+        in collectionView: UICollectionView,
+        at indexPath: IndexPath
+    ) -> UICollectionReusableView?
+}
```
이부분이 그럼 나중에 다른 화면도 섹션을 사용하면 이렇게 사용하면 될까요?
protocol 명칭으로는 다른 섹션 부분에서도 protocol을 채택하면 될까 궁금해요!

만약 섹션을 쓰는 다른곳에서도 하게 된다면 파일 위치를 이동해도 좋을 것 같습니다!!

rafa-e1 replied:
좋은 질문 및 의견 감사합니다!
SectionController는 UICollectionView 기반의 섹션 UI를 분리하고 모듈화하기 위한 프로토콜로 설계된 구조이기 때문에 다른 화면에서도 재사용이 가능합니다!
말씀주신 것처럼 이 프로토콜이 특정 화면에 종속되지 않고 공통 역할을 수행한다면 `Presentation/Common/Section` 또는 `Common/Protocol`과 같은 경로로 파일 위치를 옮기는데 더 좋은 방향이네요! 반영하도록 하겠습니다:)

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Layout/Section/Controller/VoteSectionController.swift:
```diff
@@ -0,0 +1,41 @@
+
+final class VoteSectionController: SectionController {
+    
+    func numberOfItems() -> Int { 1 }
```
```swift
    func numberOfItems() -> Int {
        section.numberOfItems
    }
```

이부분은 이렇게 안하신 이유가 궁금합니다!

rafa-e1 replied:
앗 이 부분은 제가 놓친 부분이네요. 꼼꼼하게 봐주셔서 감사합니다!

alstjr7437 reviewed (commented):
코드 줄을 보니 정말 많더라구요!!
그래도 어느정도 다 읽어보면서 리뷰 남겼습니다!!

따로 궁금한점 질문이나 사용했던걸 구경했어요 사실 ㅋㅋ...
너무 많이 들어오다보니 아직 자세하게 이해는 못했지만 사용해보면서 같이 맞춰나가요!!


혹시 다음 PR부터는 만약 

홈 투표 UI 및 내비를 만든다고 하면 중간에 작업이 뻗어 나가는걸 PR로 나눠주실 수 있을까요!
- 홈 투표 UI 및 내비게이션 바 구현
    - 	UI 구현에 필요한 Foundation 구현
    - BaseView 등 기본적인 필요한 class 구현
    - Footer View 구현
    - Header View 구현
    - ...

이런식으로 뻗어나가는게 좋을 것 같은데 어떻게 생각하시나요!?

위와 같이 하면 각 부분에 맞는 코드들만 볼 수 있어서 리뷰하기 좋고 기능 분리도 좋을 것 같습니다!

rafa-e1 commented on MatzipBook/MatzipBook/Presentation/Home/Layout/Factory/HomeLayoutFactory.swift:
```diff
@@ -0,0 +1,18 @@
+        switch section {
+        case .vote: VoteSectionLayout().layout()
+        default: DefaultSectionLayout().layout()
+        }
```
`switch` 문에서 `return` 키워드를 작성하지 않아 `Void`를 반환하게 된 점은 놓친 부분이 맞습니다. 리뷰 주신 대로 `return` 구문을 작성하여 컴파일 오류를 해결하겠습니다. 

현재는 섹션의 수가 적고 레이아웃의 복잡도가 높지 않아 초기 단계에서는 가독성과 구현 단순성을 우선시했습니다. 이후 섹션 수가 많아지고 레이아웃 생성 비용이 높아질 경우를 대비해서 `static` 프로퍼티를 활용하여 리팩터리;ㅇ하겠습니다.

rafa-e1 commented on MatzipBook/MatzipBook/Presentation/Home/Component/Footer/HomeSectionFooterView.swift:
```diff
@@ -0,0 +1,90 @@
+
+    private let seeAllButton: UIButton = UIButton(type: .system)
+
+    var onTap: (() -> Void)?
+    
+    // MARK: - Initializer
+    
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        
+        setAddTargets()
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: - Actions
+    
+    @objc private func didTapButton() {
+        onTap?()
+    }
```
감사합니다. 제안해주신 대로 `prepareForReuse()`를 구현하여 `onTap` 클로저를 해제하고 버튼 상태를 초기화하도록 수정하겠습니다.

rafa-e1 commented:
> 코드 줄을 보니 정말 많더라구요!! 그래도 어느정도 다 읽어보면서 리뷰 남겼습니다!!
> 
> 따로 궁금한점 질문이나 사용했던걸 구경했어요 사실 ㅋㅋ... 너무 많이 들어오다보니 아직 자세하게 이해는 못했지만 사용해보면서 같이 맞춰나가요!!
> 
> 혹시 다음 PR부터는 만약
> 
> 홈 투표 UI 및 내비를 만든다고 하면 중간에 작업이 뻗어 나가는걸 PR로 나눠주실 수 있을까요!
> 
> * 홈 투표 UI 및 내비게이션 바 구현
>   
>   * UI 구현에 필요한 Foundation 구현
>   * BaseView 등 기본적인 필요한 class 구현
>   * Footer View 구현
>   * Header View 구현
>   * ...
> 
> 이런식으로 뻗어나가는게 좋을 것 같은데 어떻게 생각하시나요!?
> 
> 위와 같이 하면 각 부분에 맞는 코드들만 볼 수 있어서 리뷰하기 좋고 기능 분리도 좋을 것 같습니다!

좋은 제안 감사합니다! 다음 PR부터는 기능 단위로 더 명확하게 나눠서 작업을 진행하겠습니다:)

rafa-e1 committed:
rename: 더 직관적인 이름으로 파일명 및 enum 수정

rafa-e1 committed:
feat: 홈 화면 섹션별 셀 configure 메서드 기본 구현 추가

rafa-e1 committed:
refactor: 하드 코딩된 데이터 제거 및 configureVoteButton 함수 옵셔널 이미지 지원하도록 개선

rafa-e1 committed:
feat: 홈 화면 셀 뷰모델 구현 (추천, 숨은맛집, 주변랭킹, 투표)

- HiddenGemCellViewModel, NearbyRankingCellViewModel, RecommendationCellViewModel, VoteCellViewModel 추가
- 도메인 모델을 뷰 표현 데이터로 변환하도록 설계
- 향후 API 연동 시 확장 가능하도록 초기 구조 마련

rafa-e1 committed:
rename: Recommendation을 PersonalRecommendation으로 변경하여 개인화 의미를 명확히 함

rafa-e1 committed:
feat: 도메인 모델 및 추천 섹션 타입 enum 구현

rafa-e1 committed:
remove: 미사용 코드 제거

rafa-e1 committed:
feat: 홈 섹션별 SectionController 구현 및 ViewModel 연결

rafa-e1 committed:
rename: 각 프로토콜의 역할과 의도를 명확히 표현하도록 이름 변경

rafa-e1 committed:
refactor: 폴더 구조 재정비

rafa-e1 committed:
feat: HomeSectionType에 섹션 타입 추가 및 레이아웃 메서드 구현

rafa-e1 committed:
feat: 각 섹션 더미 데이터 구성 로직 추가

rafa-e1 committed:
feat: HeaderFooterDisplayable 프로토콜 정의하여 섹션 헤더, 푸터 지원

rafa-e1 committed:
feat: 섹션별 레이아웃 분기 처리

rafa-e1 committed:
refactor: 간결하고 직관적인 이름으로 수정 및 적용

rafa-e1 committed:
refactor: HomeViewModel 프로토콜 제거 및 sections 변수명 개선

- HomeViewModelProtocol이 불필요한 추상화로 판단되어 제거
- ViewModel을 교체하거나 다형성을 활용할 요구사항이 없고
  구체 타입으로 충분히 역할을 수행 가능하므로 단순화
- 코드 가독성과 유지보수성을 위해 인터페이스 중복을 제거

rafa-e1 committed:
refactor: 미사용 코드 제거 및 변경된 네이밍 적용

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Component/Cell/VoteCollectionViewCell.swift:
```diff
@@ -0,0 +1,224 @@
+    private let titleLabel: UILabel = UILabel()
+    private let subtitleLabel: UILabel = UILabel()
+    
+    private let leftNameImageView: UIImageView = UIImageView()
+    private let leftNameLabel: UILabel = UILabel()
+    
+    private let rightNameImageView: UIImageView = UIImageView()
+    private let rightNameLabel: UILabel = UILabel()
```
조금 사소한 부분이지만 해당 부분들을 leftItem과 rightItem으로 ImageView, Label, Button을 컴포넌트로 묶을 수도 있겠네요!

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/HomeSection/Controller/HiddenGemSectionController.swift:
```diff
@@ -0,0 +1,83 @@
+            return nil
+        }
+
+        header.configure(title: headerTitle)
```
header부분의 title이 nil일때 보여주는 텍스트는 필요 없을까요?

rafa-e1 replied:
현재 구조상 NearbyRankingSection, PersonalRecommendationSection, HiddenGemSection은 항상 헤더 타이틀이 존재하고 VoteSection처럼 헤더, 푸터가 필요 없는 섹션만 nil을 전달하고 있습니다! 그래서 headerTitle이 nil인 경우는 의도적으로 헤더가 필요 없는 섹션이기 때문에 별도의 기본값은 적용하지 않아도 괜찮다고 판단했습니다. 이후에 headerTitle이 optional이면서도 placeholder를 보여줘야 하는 상황이 생기면 그때 반영해도 충분할 것 같습니다!

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/HomeSection/Controller/NearbyRankingSectionController.swift:
```diff
@@ -0,0 +1,83 @@
+        }
+
+        let cellData: NearbyRanking = items[indexPath.item]
+        let viewModel: NearbyRankingCellViewModel = NearbyRankingCellViewModel(
+            nearbyRanking: cellData
+        )
```
아래 VC도 같은 부분인데 cellForItem을 할때마다 ViewModel을 init하는게 맞나요??
따로 인스턴스를 가지고 있지 않고 만드는 부분이 궁금합니다!

rafa-e1 replied:
좋은 질문 감사합니다!

현재 구조에서는 cellForItem에서 매번 NearbyRankingCellViewModel을 생성하고 있는데 뷰모델이 단순히 데이터 변환만 담당하고 있어서 상태를 보존하거나 복잡한 로직을 처리하지 않기 때문에 매번 생성해도 성능상 큰 문제가 없다고 판단했습니다! 이후에 Diffable DataSource를 도입해소 섹션과 아이템 단위의 상태를 더 효율적으로 관리할 예정이라 뷰모델을 따로 캐싱하거나 배열로 관리하는 방식으로 개선할 여지도 고려하고 있습니다. 즉, 현재 구조는 간단성을 우선한 설계이고 데이터 업데이트나 셀 상태 관리가 복잡해질 때 Diffable과 함께 뷰모델의 재사용을 적용할 계획입니다.

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/HomeSection/Enum/HomeSectionType.swift:
```diff
@@ -0,0 +1,22 @@
+
+import UIKit
+
+enum HomeSectionType: Int {
+    case vote, nearbyRanking, personalRecommendation, hiddenGem
+}
+
+extension HomeSectionType {
+    
+    func sectionLayout() -> NSCollectionLayoutSection {
+        switch self {
+        case .vote: return HomeLayout.voteSectionLayout
+        default: return HomeLayout.defaultSectionLayout
+        }
+    }
+}
```
Home Section을 enum 나눠서 가지고 있으니 어떠한 섹션이 있는지 보기도 좋고 기능적으로 확장성도 좋네요!

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/HomeSection/Layout/HomeSectionLayoutProvider.swift:
```diff
@@ -0,0 +1,23 @@
+    static let defaultSectionLayout: NSCollectionLayoutSection = DefaultSectionLayout().layout()
+}
+
+final class HomeSectionLayoutProvider: HomeSectionLayoutProviding {
+    
+    func layout(for section: HomeSectionType) -> NSCollectionLayoutSection {
+        switch section {
+        case .vote: return HomeLayout.voteSectionLayout
+        default: return HomeLayout.defaultSectionLayout
+        }
+    }
+}
```
섹션 팩토리랑 layout을 나눈 이유는
View Layout잡는 코드와 내용이 들어가는 코드를 분할하기 위해서 일까요?

rafa-e1 replied:
섹션에 채워질 데이터 구성 책임과 그 데이터를 어떻게 레이아웃으로 표현할지 책임을 분리하여 관심사를 구분했습니다!

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Model/NearbyRanking.swift:
```diff
@@ -0,0 +1,22 @@
+
+import UIKit
+
+struct NearbyRanking {
+    var ranking: Int
+    let name: String
+    let image: UIImage?
+    var rating: Double
+    var distance: Double
+    let foodCategory: String
+    let themeCategory: String
```
해당 레스토랑의 고유 아이디가 없어도 될지 
storeId가 검색에는 있는데 맛집 API명세에는 없어서 서버 분들과도 이야기해봐야겠네요

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/Model/Vote.swift:
```diff
@@ -0,0 +1,14 @@
+struct Vote {
+    let leftRestaurant: Restaurant
+    let rightRestaurant: Restaurant
+    let remainingTime: Int
```
이부분은 따로 Date로 가지지 않고 Int로 가지고 있는걸까요?

rafa-e1 replied:
아직 API 명세가 완성되지 않아서 우선은 UI에서 바로 표시할 수 있도록 임의로 Int 타입으로 남은 시간을 설정해두었습니다! 이후에 API 명세가 확정되면 서버에서 전달되는 데이터 형식에 맞춰 Date 혹은 다른 적절한 타입으로 수정할 계획입니다:)

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/ViewModel/HomeViewModel.swift:
```diff
@@ -0,0 +1,30 @@
+
+final class HomeViewModel {
+    
+    var universityName: String = "부경대학교"
```
이친구는 따로 private(set)으로 안한 이유가 있을까요?!

rafa-e1 replied:
이후에 대학을 변경할 수 있도록 만들 계획이기 때문입니다!

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/ViewModel/NearbyRankingCellViewModel.swift:
```diff
@@ -0,0 +1,34 @@
+import UIKit
+
+final class NearbyRankingCellViewModel {
+    
+    private let nearbyRanking: NearbyRanking
+    
+    init(nearbyRanking: NearbyRanking) {
+        self.nearbyRanking = nearbyRanking
+    }
+    
+    var ranking: Int { nearbyRanking.ranking }
+    var restaurantName: String { nearbyRanking.name }
+    var restaurantImage: UIImage? { nearbyRanking.image }
+    var rating: Double { nearbyRanking.rating }
+    var starCount: Int { nearbyRanking.starCount }
+    var formattedDistance: String {
```
어짜피 let이라면 
var 로 아래 인스턴스를 뽑아내지 않고 아래와 같이 안하신 이유가 있을까요?!

```suggestion
    
   let nearbyRanking: NearbyRanking
    
    init(nearbyRanking: NearbyRanking) {
        self.nearbyRanking = nearbyRanking
    }
    
    var formattedDistance: String {
```

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/ViewModel/PersonalRecommendationCellViewModel.swift:
```diff
@@ -0,0 +1,24 @@
+import UIKit
+
+final class PersonalRecommendationCellViewModel {
+    
+    private let personalRecommendation: RecommendedRestaurant
+    
+    init(personalRecommendation: RecommendedRestaurant) {
+        self.personalRecommendation = personalRecommendation
+    }
+    
+    var restaurantImage: UIImage? { personalRecommendation.image }
+    var restaurantRating: Double { personalRecommendation.rating }
+    var restaurantName: String { personalRecommendation.name }
+    var restaurantDescription: String { personalRecommendation.description }
+    var foodCategory: String { personalRecommendation.foodCategory }
+    var themeCategory: String { personalRecommendation.themeCategory }
```
해당 부분도 위와 같이 궁금해요!

alstjr7437 commented on MatzipBook/MatzipBook/Presentation/Home/ViewModel/VoteCellViewModel.swift:
```diff
@@ -0,0 +1,27 @@
+        self.vote = vote
+    }
+    
+    var leftRestaurantName: String { vote.leftRestaurant.name }
+    var leftRestaurantImage: UIImage? {
+        vote.leftRestaurant.thumbnail ?? .imgDummy0
+    }
+    var rightRestaurantName: String { vote.rightRestaurant.name }
+    var rightRestaurantImage: UIImage? {
+        vote.rightRestaurant.thumbnail ?? .imgDummy1
+    }
+    var remainingTime: Int { vote.remainingTime }
```
이 부분도!🙇

rafa-e1 replied:
현재는 지적해주신 뷰모델이 모델의 데이터를 감싼 형태로 설계되어 있습니다.
향후 Cell의 표현 방식(예: 거리 단위 포맷, 별점 표시, 카테고리 텍스트 포맷 등)이 바뀔 때 뷰모델에서 쉽게 가공할 수 있도록 하고 UI 레이어에서의 변경에 따라 모델을 직접 건드리지 않도록 명확히 역할을 나누기 위해 뷰모델을 분리했습니다!

말씀주신대로 단순히 let 프로퍼티를 직접 넘겨도 되지만 표현 책임을 뷰모델에 위임해두면 추후 UI 변경이 발생해도 모델에는 영향이 없어서 유지보수가 더 편할 것으로 판단했습니다.

나중에 뷰모델이 너무 얇으면 제거하고 바로 모델을 전달하는 방식으로도 변경 가능하니 의견 주시면 조율해 보겠습니다!

alstjr7437 reviewed (approved):
아직 아무것도 안하고 거의 리뷰만 달고 있는 저가 뭔가 정말 죄송하네요...

매번 작업은 안해놓고 리뷰만 달아서 정말 죄송합니다,, ㅠㅠ

저가 할거리만 늘리는 것 같네요 ㅠ,, 이번에 컨벤션이 갖춰지면 열심히 로그인, 온보딩 화면 작업할게요
